File: 04-05/calculadora.py
def funcSucc(x):
    return x + 1

def funcPred(x):
    return x - 1

def soma(a, b):
    if (b == 0):
        return a
    elif b > 0: 
        return soma(funcSucc(a), funcPred(b))
    elif b < 0:
        return soma(funcPred(a), funcSucc(b))

def sub(a, b):
    return soma(a, -b)

def potencia(a, b):
    if b == 0:
        return 1
    elif b > 0: 
        return a * potencia(a, b-1)

def calculadora():
    print("""
    Escolha uma operacao
    0 - Sair
    1 - Adicao
    2 - Subtracao
    3 - Potenciacao
 """)
    op = int(input())
    if op == 0:
        print("Saindo...")
    else:
        if op == 1:
            print("SOMA")
            x = int(input("Digite o primeiro numero: "))
            y = int(input("Digite o segundo numero: "))
            print("Resultado: ", soma(x, y))
        elif op == 2:
            print("SUBTRACAO")
            x = int(input("Digite o primeiro numero: "))
            y = int(input("Digite o segundo numero: "))
            print("Resultado: ", sub(x, y))
        elif op == 3:
            print("POTENCIACAO")
            x = int(input("Digite o primeiro numero: "))
            y = int(input("Digite o segundo numero: "))
            print("Resultado: ", potencia(x, y))
        else:
            print("Operação Invalida")
        calculadora()

if __name__ == "__main__":
    calculadora()

File: 04-05/calculadora.hs
funcSucc :: Int -> Int
funcSucc x = x + 1

funcPred :: Int -> Int
funcPred x = x - 1

add a b 
    | b == 0 = a
    | b > 0 = add(funcSucc a) (funcPred b)
    | b < 0 = add(funcPred a) (funcSucc b)

sub a b = add a (-b)

potencia a b
    | b == 0 = 1
    | b > 0 = a * (potencia a (b - 1))

calculadora = do
    putStrLn "Selecione uma operação:"
    putStrLn "0 - Sair"
    putStrLn "1 - Adição"
    putStrLn "2 - Subtração"
    putStrLn "3 - Potenciação"
    op <- readLn
    if op == 0 then
        putStrLn("Saindo...")
    else do
        if op == 1 then do
            putStrLn("SOMA")
            putStrLn("Digite o primeiro número:")
            num1 <- readLn
            putStrLn("Digite o segundo número:")
            num2 <- readLn
            putStrLn ("Resultado: " ++ show (add num1 num2))
        else if op == 2 then do
            putStrLn("SUBTRAÇÃO")
            putStrLn("Digite o primeiro número:")
            num1 <- readLn
            putStrLn("Digite o segundo número:")
            num2 <- readLn
            putStrLn ("Resultado: " ++ show (sub num1 num2))
        else if op == 3 then do
            putStrLn("POTENCIAÇÃO")
            putStrLn("Digite a base:")
            base <- readLn
            putStrLn("Digite o expoente:")
            expoente <- readLn
            putStrLn ("Resultado: " ++ show (potencia base expoente))
        else
            putStrLn("Operacao invalida")
        calculadora

main :: IO ()
main = do
    calculadora